jax/core.py

```diff
@@ -309,6 +309,7 @@ def full_raise(self, val):
 
 
   def pure(self, val):
+    """Given a concrete value, makes a Tracer for it."""
```

Comment:
There's no reason for `val` to be concrete (its abstract value could be at any level), and it may already be a Tracer; it's just not lifted into this trace. It may be better to delete this docstring.

Reply:
I have deleted it, but I do think we need this to be documented eventually.

Reply:
From what I can see in the code, trace.pure is only called in [trace.full_raise](https://github.com/google/jax/blob/18420936c4ce5aca7349766de66c21ea8efdec53/jax/core.py#L284) and only if the value is not a Tracer. 

I think that the Trace class is very important in the code, and it is worth documenting how it works, and I think that the best place is as docstrings in the `Trace` super class.